Add instructions on creating a stack in the operators guide (#316)
buildpacks/docs: content/docs/operator-guide/create-a-stack.md

+++
title="Create a stack"
weight=2
+++

Creating a custom [stack][stack] allows you to configure the base images for the build-time environment for your [builder][builder] and the run-time for your application.

<!--more-->

## Prerequisites

Before we get started, make sure you've got the following installed: 

{{< download-button href="https://store.docker.com/search?type=edition&offering=community" color="blue" >}} Install Docker {{</>}}

## Creating a custom stack

> In this tutorial we will create a sample stack based on `Ubuntu Bionic`. To create a custom stack, we simply need to create customized build and run images. Let's see how we can do so!

Reviewer: ```suggestion
In this tutorial we will create a sample stack based on `Ubuntu Bionic`. To create a custom stack, we need to create customized build and run images. Let's see how we can do so!
```

### Create a common base image

Let's start by creating a base image containing layers that will be required by both the `build` and `run` images. In order to do this, switch to a clean workspace and create a `Dockerfile` as specified below:

#### Defining the base

> We start with choosing `ubuntu:bionic` as our `base` image. Since we will be reusing these layers in both our build and run images we will be defining a common base image and leveraging [Docker's multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) to ensure this acts as the common base image for both our build-time and run-time environment.

Reviewer: ```suggestion
We start with `ubuntu:bionic` as our `base` image. Since we will be reusing these layers in both our build and run images we will be defining a common base image and leveraging [Docker's multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) to ensure this acts as the common base image for both our build-time and run-time environment.
```

```Dockerfile
# 1. Set a common base
FROM ubuntu:bionic as base
```

#### Set required CNB information

> Next up we will be setting up the base image as required by the [Cloud-Native Buildpack specification][stack-spec] noted below.

Reviewer: ```suggestion
Next, we will be setting up the base image as required by the [Cloud-Native Buildpack specification][stack-spec] noted below.
```

Reviewer: just flows a little better with out the double "up"

##### Specification

**Labels**

| Name                     | Description              | Format |
| ------------------------ | ------------------------ | ------ |
| `io.buildpacks.stack.id` | Identifier for the stack | String |

**Environment Variables**

| Name           | Description                            |
| -------------- | -------------------------------------- |
| `CNB_STACK_ID` | Identifier for the stack               |
| `CNB_USER_ID`  | UID of the user specified in the image |
| `CNB_GROUP_ID` | GID of the user specified in the image |
<p class="spacer"></p>

> **NOTE:** The **stack identifier** implies compatibility with other stacks of that same identifier. For instance, a custom stack may use
> `io.buildpacks.stacks.bionic` as its identifier so long as it will work with buildpacks that declare compatibility with the
> `io.buildpacks.stacks.bionic` stack.

> The `CNB_USER_ID` is the `UID`  of the user as which the `detect` and `build` steps are are run. The given user **MUST** be a non-root user

Reviewer: ```suggestion
The `CNB_USER_ID` is the `UID`  of the user as which the `detect` and `build` steps are run. The given user **MUST NOT** be a root user
```

and have it's home directly writeable. `CNB_GROUP_ID` is the primary `GID` of the above user.

Let's update the `Dockerfile` to reflect the above specification.

```Dockerfile
# 1. Set a common base
FROM ubuntu:bionic as base

# ========== ADDED ===========
# 2. Set required CNB information
ARG cnb_uid=1000

> ARG cnb_gid=1000

Reviewer: Are the `ARG`s necessary? If not, we could save a few lines by omitting them. IDK, it's fine to keep them though.

Author: Removed them.

ARG stack_id="io.buildpacks.samples.stacks.bionic"
ENV CNB_USER_ID=${cnb_uid}
ENV CNB_GROUP_ID=${cnb_gid}
ENV CNB_STACK_ID=${stack_id}
LABEL io.buildpacks.stack.id=${stack_id}

# 3. Create the user
RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb
```

#### Install common packages

> Next up, we will be installing any common packages that we want to make available to both our build-time and run-time environment. Our `Dockerfile` should now look like -

Reviewer: ```suggestion
Next up, we will be installing any system packages that we want to make available to both our build-time and run-time environment. Our `Dockerfile` should now look like -
```

Reviewer: "common" felt odd to me. but it's just me

```Dockerfile
# 1. Set a common base
FROM ubuntu:bionic as base

# 2. Set required CNB information
ARG cnb_uid=1000
ARG cnb_gid=1000
ARG stack_id="io.buildpacks.samples.stacks.bionic"
ENV CNB_USER_ID=${cnb_uid}
ENV CNB_GROUP_ID=${cnb_gid}
ENV CNB_STACK_ID=${stack_id}
LABEL io.buildpacks.stack.id=${stack_id}

# 3. Create the user
RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

# ========== ADDED ===========
# 4. Install common packages
RUN apt-get update && \
  apt-get install -y xz-utils ca-certificates && \
  rm -rf /var/lib/apt/lists/*
```

That should be it for our base image! Let's verify that we can successfully build this image by running:

```bash
docker build . -t cnbs/sample-stack-base:bionic --target base
```

### Creating the run image

> Next up, we will be creating the run image. The run image is the base image for your runtime application environemnt.

Reviewer: ```suggestion
Next up, we will create the run image. The run image is the base image for your runtime application environemnt.
```

In order to create our run image all we need to do is to set the run image's `USER` to the user with `CNB_USER_ID`. Our final `Dockerfile` for the build image should look like - 

```Dockerfile
# 1. Set a common base
FROM ubuntu:bionic as base

# 2. Set required CNB information
ARG cnb_uid=1000
ARG cnb_gid=1000
ARG stack_id="io.buildpacks.samples.stacks.bionic"
ENV CNB_USER_ID=${cnb_uid}
ENV CNB_GROUP_ID=${cnb_gid}
ENV CNB_STACK_ID=${stack_id}
LABEL io.buildpacks.stack.id=${stack_id}

# 3. Create the user
RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

# 4. Install common packages
RUN apt-get update && \
  apt-get install -y xz-utils ca-certificates && \
  rm -rf /var/lib/apt/lists/*

# ========== ADDED ===========
# 5. Start a new run stage
FROM base as run

# 6. Set user and group (as declared in base image)
USER ${CNB_USER_ID}:${CNB_GROUP_ID}
```

That should be it for our run image! Let's verify that we can successfully build this image by running:

```bash
docker build . -t cnbs/sample-stack-run:bionic --target run
```

### Creating the build image

> Next up, we will be creating the build image. The build image is the base image for you builder and should contain any common build-time

Reviewer: ```suggestion
Next up, we will create the build image. The build image is the base image for you builder and should contain any common build-time
```

+++
title="Create a stack"
weight=2
+++

Creating a custom [stack][stack] allows you to configure the base images for the build-time environment for your [builder][builder] and the run-time for your application.

<!--more-->

## Prerequisites

Before we get started, make sure you've got the following installed: 

{{< download-button href="https://store.docker.com/search?type=edition&offering=community" color="blue" >}} Install Docker {{</>}}

## Creating a custom stack

In this tutorial we will create a sample stack based on `Ubuntu Bionic`. To create a custom stack, we simply need to create customized build and run images. Let's see how we can do so!

### Create a common base image

Let's start by creating a base image containing layers that will be required by both the `build` and `run` images. In order to do this, switch to a clean workspace and create a `Dockerfile` as specified below:

#### Defining the base

> We start with choosing `ubuntu:bionic` as our `base` image. Since we will be reusing these layers in both our build and run images we will be defining a common base image and leveraging [Docker's multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) to ensure this acts as the common base image for both our build-time and run-time environment.

Author: I used a multi-stage build to ensure consistency in the build/run images and having it all in one file.

```

### Creating the build image

Next up, we will be creating the build image. The build image is the base image for you builder and should contain any common build-time
environment required by your buildpacks.

#### Install build packages

Let's modify the `Dockerfile` to look like - 

```Dockerfile
# 1. Set a common base
FROM ubuntu:bionic as base

# 2. Set required CNB information
ARG cnb_uid=1000
ARG cnb_gid=1000
ARG stack_id="io.buildpacks.samples.stacks.bionic"
ENV CNB_USER_ID=${cnb_uid}
ENV CNB_GROUP_ID=${cnb_gid}
ENV CNB_STACK_ID=${stack_id}
LABEL io.buildpacks.stack.id=${stack_id}

# 3. Create the user
RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

# 4. Install common packages
RUN apt-get update && \
  apt-get install -y xz-utils ca-certificates && \
  rm -rf /var/lib/apt/lists/*

# ========== ADDED ===========
# 5. Start a new run stage
FROM base as run

# 6. Set user and group (as declared in base image)
USER ${CNB_USER_ID}:${CNB_GROUP_ID}

# ========== ADDED ===========
# 7. Start a new build stage
FROM base as build

# 8. Install packages that we want to make available at build time

> RUN apt-get update && \

Author: Modified this a little from our samples repo to make it self contained and avoid using COPY
